Upgrade from woefully outdated jQuery version

RafaelJohn9 opened the pull request:
## Changes

<!-- List the changes this PR makes. -->

- fix #2004

## Context

<!-- Explain why you're making these changes. -->

jQuery has not been meaningfully updated in over a decade. The previous version in use , **jQuery 1.7.1 (released in 2011)** , predates widespread support for modern JavaScript APIs. It was last updated in this project around 2013, and since then, no further upgrades were made despite **over 50 releases** and critical improvements in the jQuery ecosystem.

This long-standing neglect created growing technical debt:
- Increasing incompatibility with modern development tools
- Accumulation of deprecated patterns (`.bind()`, `.click()`)
- Potential security and performance concerns
- Poor developer experience when debugging or extending JS

While the site continued to function thanks to jQuery Migrate and browser backward compatibility, this was **not sustainable maintenance** .

---

## Why Upgrade Now?

We are modernizing the JavaScript foundation to:
- [x] **Improve compatibility** with current and future browser standards
- [x] **Leverage bug fixes and performance improvements** from jQuery 3.x
- [x] **Enable use of modern jQuery features** (e.g. improved `.on()`, better event delegation)
- [x] **Support upcoming enhancements** that depend on reliable, up-to-date tooling
- [x] **Remove reliance on deprecated APIs** that hinder maintainability
- [x] **Align with security best practices** by using actively supported versions

jQuery 3.x dropped support for legacy browsers (IE 6–8), which we no longer need to support. This allows us to benefit from a leaner, faster, and more predictable codebase.

---

## Migration Strategy

This upgrade was done incrementally to ensure stability:
1. **Added jQuery Migrate 1.2.1** to detect deprecations
2. **Refactored legacy event bindings** (`.click()` → `.on()`, `.bind()` → `.on()`)
3. **Upgraded to jQuery 3.7.1 + jQuery UI 1.14.1** with modern Migrate
4. **Verified functionality** across all interactive components
5. **Removed Migrate** after confirming no deprecation warnings
6. **Cleaned up outdated asset comments** in `application.js`

All changes preserve existing behavior. The result is a safer, more maintainable JavaScript foundation.

----

This work closes a long-overdue maintenance gap and sets the stage for more confident, modern front-end development.

dscho commented:
You will most likely have to follow [this guidance](https://jquery.com/upgrade-guide/), in particular using [`jquery-migrate`](https://github.com/jquery/jquery-migrate).

Also, the structure of these commits is _chronological_, in dear need of becoming _logical_ instead. Keep an eye on reviewability. Hiding the change that drops `jquery.defaultvalue` in a different commit is no good, in particular when the commit message does not even explain nor provide references justifying this drop.

RafaelJohn9 commented:
hey @dscho , 

on `jquery.defaultValue` its seems that it hasn't been  maintained for a while:  https://github.com/cloudchen/jquery.defaultValue

dscho commented:
> on `jquery.defaultValue` its seems that it hasn't been maintained for a while: https://github.com/cloudchen/jquery.defaultValue

That's excellent information, which dearly wants to go into the commit message.

dscho commented on assets/js/application.js:
```diff
@@ -565,7 +564,7 @@ var Downloads = {
 
   observePopState: function() {
-    onPopState(function() {
+    onPopState(function () {
```
That's not updating a deprecated function call, but it changes the code style to disagree with the remainder of the file.

dscho commented on assets/js/application.js:
```diff
@@ -43,8 +43,7 @@ $(document).ready(function() {
   if (window.history && window.history.pushState) {
-    return $(window).bind('popstate', function(event) {
-      var section;
+    return $(window).on('popstate', function () {
```
Can you find some documentation about this? If so, that would be good information for the commit message.

Also: Does this new call work with the outdated JQuery version that `git-scm.com` currently has? If so, this commit needs to come _before_ the upgrade to the newer JQuery.

RafaelJohn9 replied:
this was what was suggested on `jquery migrate` during migration

dscho replied:
That was not at all clear to me from reading the commit message and the diff. And chances are that you won't remember in six months from now, unless you record that vital information in the commit message, with a bit more verbosity than "this was what was suggested by `jquery migrate`"...

dscho commented on assets/js/application.js:
```diff
@@ -780,12 +779,12 @@ var PostelizeAnchor = {
     : $('#scrollToTop').fadeOut();
 });
-$('#scrollToTop').click(function(e) {
+$('#scrollToTop').on('click', function (e) {
```
I see a lot of `$(...).<something>(...)` -> `$(...).on('<something>, ...)` changes here. Do they all belong to the same class of deprecation, logically? If so, it's fine to cluster them in one large commit. Otherwise it makes reviewing easier (and hence bug catching, too, which is what reviewing is all about) to split the changes up by logical topic (which would then need to be described in the commit message, accompanied by references you can find).

dscho replied:
Okay, so now there is proper documentation about this. How can we be certain that all of the deprecated (and probably removed from 3.x?) stuff has been replaced, i.e. that nothing was missed?

I have a _suspicion_ that the [jQuery Migrate plugin](https://github.com/jquery/jquery-migrate) is actually highly desirable in this context, something I might have mentioned before but I cannot see has so far resulted in any action...

dscho commented on assets/js/application.js:
```diff
@@ -10,8 +10,8 @@
 //
-//= require jquery-1.7.1.min
-//= require jquery-ui-1.8.18.custom.min
+//= require jquery-3.7.1.min
+//= require jquery-ui-1.14.1.min
```
It does not make much sense to split the JQuery version updates in `application.js` from the ones in the header.

I am glad that you dropped the unrelated change from this commit where `jquery.defaultvalue` was dropped. But it seems you no longer drop this altogether? Is it still supported? Does it support the newest JQuery version?

RafaelJohn9 replied:
@dscho , so I should basically cluster:

1. changes in `application.js`
2. changes in `footer.html`
3. the new files at `static/js`

dscho replied:
Okay, so now this is clustered properly, but the commit message needs _a lot_ of love.

Please do mention from where you obtained the `.min.js` files. And wouldn't it be a lot better to rely on jQuery's own CDN at https://releases.jquery.com/? Then we would not have to carry copies of the files _and_ you would automatically ascertain that the jQuery code has not been tampered with.

dscho reviewed (changes requested):
I am still convinced that you will most likely have to follow [this guidance](https://jquery.com/upgrade-guide/), in particular using [jquery-migrate](https://github.com/jquery/jquery-migrate).

Either you did (but didn't talk about it in the commit messages, which you should, though), or you didn't (but you should).

Speaking of commit messages: For something as consequential as a JQuery update from an ancient version to a current one, you definitely need to up your commit message writing fu. 
Please follow the guidance in https://github.blog/2022-06-30-write-better-commits-build-better-projects/ to improve it, in particular with a strong focus on this part:

  | What you’re doing | Why you’re doing it
-- | -- | --
High-level (strategic) | **Intent** (what does this accomplish?) | **Context** (why does the code do what it does now?)
Low-level (tactical) | **Implementation** (what did you do to accomplish your goal?) | **Justification** (why is this change being made?)

dscho commented:
> > on `jquery.defaultValue` its seems that it hasn't been maintained for a while: https://github.com/cloudchen/jquery.defaultValue
> 
> That's excellent information, which dearly wants to go into the commit message.

Okay, after mulling it over a bit and `git grep -i defaultvalue`-ing, I come to a totally different conclusion: `jquery.defaultValue` has not been _updated_ in quite a while. But that's because it is perfect, not because it isn't maintained. It does its job as it's supposed to do, and before you remove anything like that, you need to make sure that it is not used (otherwise your PR needs to be rejected out of hand). And it [is used](https://github.com/git/git-scm.com/blob/a6a37a364b947db4c173d7979da076f694033ffa/assets/js/application.js#L100)...

dscho commented on data/guis/giggle.yml:
```diff
@@ -1,5 +1,5 @@
 ---
-name: "giggle"
+name: "giggle (no longer under active development)"
```
What does this change do in the PR to "Refactor/update/jquery"? (Which, by the way, is a much improvable description for what you are trying to accomplish.)

RafaelJohn9 replied:
hey @dscho ,

```

commit 9381210304dcaa02f1a3836134beb9683e12fff9
Author: Jonathan Ballet <jon@multani.info>
Date:   Sat Jun 21 18:58:42 2025 +0200

    GUI: giggle is not longer in active development
    
    The website points to https://gitlab.gnome.org/Archive/giggle, which is marked as "archived" ; [latest tag 0.7 is 12 years old](https://gitlab.gnome.org/Archive/giggle/-/tags) now.


```

I'm not the one who sneaked that commit in

dscho replied:
@RafaelJohn9 you may have rebased it, [it has you as committer](https://github.com/git/git-scm.com/commit/9381210304dcaa02f1a3836134beb9683e12fff9) on record:

<img width="527" height="96" alt="image" src="https://github.com/user-attachments/assets/1582a666-de79-41c6-973b-24968b034cef" />

RafaelJohn9 replied:
ooh, yea, my bad,

dscho reviewed (commented):
This looks already a lot better, but there is still stuff to do.

RafaelJohn9 commented:
Hey @dscho , I used Jquery Migrate plugin, though I removed it after updating the deprecated parts,

would you recommend, leaving it inside the repo even after the upgrade ?

dscho commented:
> Hey @dscho , I used Jquery Migrate plugin, though I removed it after updating the deprecated parts,
> 
> would you recommend, leaving it inside the repo even after the upgrade ?

I would not recommend leaving it inside, but I would strongly suggest to start documenting what was done.

If you use jQuery Migrate, for example, I would have expected an initial commit to add it, to describe in its commit message what that plugin does (with a helpful link to the website) and why it was used in the first place. I would have expected the next commits to do a good job describing what problems the plugin showed and how they were resolved. The final commit I would have expected to be the one to remove the jQuery Migrate Plugin, stating that its job is done now and the migration is complete. 

That would also have prevented forcing any reader to guess as to what has been done and having to pry vital information e.g. about the plugin and its use in a subsequent conversation. If I would not have asked, the details about how this PR came about would be lost. That story should have been written down in the commit messages to begin with, though. If it wasn't written down, it's as if it had never happened.

dscho commented:
Hmm. Something is telling me that this is _not_ working as expected...
<img width="1146" height="76" alt="image" src="https://github.com/user-attachments/assets/566a38a9-3a19-4825-9c67-7404d32fdc13" />

dscho commented on layouts/partials/footer.html:
```diff
@@ -11,8 +11,9 @@
 
-<script src="{{ relURL "js/jquery-1.7.1.min.js" }}"></script>
-<script src="{{ relURL "js/jquery-ui-1.8.18.custom.min.js" }}"></script>
+<script src="https://code.jquery.com/jquery-migrate-3.4.1.min.js"></script>
```
Shouldn't this be v1.4.1 _first_, to upgrade across the jQuery v1.9.x boundary?

And, as per [the documentation](https://github.com/jquery/jquery-migrate/tree/1.x-stable/#README):

> In your web page, load this plugin _after_ the script for jQuery: [...]

That problem is highly likely to be the root cause of:

<img width="1146" height="76" alt="image" src="https://github.com/user-attachments/assets/566a38a9-3a19-4825-9c67-7404d32fdc13" />

dscho replied:
@RafaelJohn9 Please integrate https://github.com/dscho/git-scm.com/commit/631599fa518ac024053a9ab15dd8725f91a53ba9.

Also, please split this change apart: Adding jQuery Migrate is, logically speaking, a _very very different_ kind of change than upgrading the jQuery version.

Also document (in the commit message) that the guidance is to upgrade in two steps, first cross the v1.9 boundary, then the v3 boundary.

Then cross them in that order, via two separate commits.

Also, please do use the opportunity to switch away from the self-hosted `jquery*.js` files to the CDN ones as recommended by the jQuery project itself.

dscho replied:
Oh, and the commit at the end that _removes_ jQuery Migrate again is missing...

RafaelJohn9 replied:
I'll first add  the v1.9  migrate then, methodically move to v3 boundary

I'll also add the last commit to remove the jQuery Migrate :handshake:

dscho replied:
Good!

Once that's done, could you reword the PR title so that it reads a bit easier? Something like "Upgrade from woefully outdated jQuery version".

To elaborate on what motivates this change, the "Context" in the PR description should not be left empty but say something about potentially wanting to use newer jQuery features, benefit from bug fixes and support for newer browsers, apart from it just being poor workmanship to neglect central components for years.

There is also enough room in that empty space to describe when jQuery was last updated, and how frequently it was updated before that time.

dscho replied:
@RafaelJohn9 since you asked me, personally, to review, I need to point out that that [the PR description](https://github.com/git/git-scm.com/pull/2022#issue-3197214366) is still in dear need of elaborating on what motiates this change, the "Context" in the PR description should not be left empty but say something about potentially wanting to use newer jQuery features, benefit from bug fixes and support for newer browsers, apart from it just being poor workmanship to neglect central components for years.

There is also enough room in that empty space to describe when jQuery was last updated, and how frequently it was updated before that time.

RafaelJohn9 replied:
My bad, forgot about that :sweat_smile: ,

thank you for your patience

dscho replied:
Thank _you_ for your persistence and stamina! I know I ask a lot.

dscho replied:
@RafaelJohn9 please ping me when it's done, I will then use it as part of the merge commit message.

RafaelJohn9 replied:
@dscho ,  improved the PR description   :handshake:

dscho replied:
Thank you! It's now part of the (immutable) commit history: https://github.com/git/git-scm.com/commit/f847b394bbd131e1f8ad22a233c0a6dd064fd809

dscho commented on assets/js/application.js:
```diff
@@ -10,8 +10,8 @@
 //
-//= require jquery-1.7.1.min
-//= require jquery-ui-1.8.18.custom.min
+//= require jquery-3.7.1.min
+//= require jquery-ui-1.14.1.min
```
The first line of the commit message talks about adding two files. That's not the only change this commit makes, though, is it? It upgrades _at the same time_.

Let's just stop vendoring in the jQuery `.js` files. There are perfectly fine CDN versions of them; As a bonus, browsers can rely on the same `.js` files across sites, they don't need to cache a separate copy _just_ for git-scm.com.

dscho reviewed (changes requested):
Great work on 76644af9da86a0f351b6f21f4a007dd3bf63ca28!

RafaelJohn9 committed:
chore: upgrade jQuery and introduce Migrate for safe modernization

Upgrade jQuery from 1.7.1 to 1.9.1 and jQuery UI from 1.8.18 to 1.9.1 to
bridge toward a modern codebase. While 1.9.1 is still outdated by today's
standards, it introduces critical deprecation warnings and improves
compatibility with newer APIs, serving as a necessary intermediate step
before a full jump to jQuery 3.x.

Importantly, this commit adds jQuery Migrate 1.2.1 from the official CDN
to detect and log deprecated usage patterns that would otherwise break
silently during future upgrades. Migrate will surface issues like:
- Calls to deprecated methods (.live(), .bind() in certain contexts)

These warnings are essential for identifying technical debt in our
existing JavaScript (e.g., in application.js)
and planning safe refactors.

This is Step 1 of a phased migration strategy:
1. Upgrade to jQuery 1.9.1 + Migrate → current commit
2. Audit console warnings and modernize code
3. Upgrade to jQuery 3.7.1 with jquery-migrate-3.4.1
4. Remove Migrate once all deprecations are resolved

The use of CDN-hosted assets simplifies version management and ensures
we’re not bundling outdated copies.

Note: Source map 404s may appear (jquery.min.map) — these are harmless
and expected for older jQuery versions. We will address them in a later
upgrade when moving to modern, actively supported releases.

RafaelJohn9 committed:
refactor: modernize event binding using .on() instead of deprecated methods

Replace all instances of direct event binding methods (.click(), .focus(),
.bind(), .blur(), etc.) with the modern .on() API across the codebase.

The jQuery .on() method has been the recommended way to attach event
handlers since jQuery 1.7 (2011), superseding:
- .bind()
- .live() (removed in 1.9)
- .delegate()
- shorthand methods (.click(), .hover()) when delegation or cleanup is needed

While the shorthand methods still work, using .on() consistently provides:
- A single, unified pattern for event attachment
- Clear intent for both direct and delegated events
- Better compatibility with future jQuery versions
- Easier event namespace management and removal (.off())

This change touches multiple modules:
- GitTurns20: popstate handling and tagline click counter
- Search: focus, blur, keydown, and result interaction
- Dropdowns: toggle activation via click
- Forms: input selection
- Downloads: OS filter navigation
- DarkMode: toggle button interaction
- ScrollToTop: scroll monitoring and click behavior

Notable improvements:
- $(window).bind('popstate', ...) → $(window).on('popstate', ...): aligns with current best practices
- $("#tagline").click(...) → .on('click', ...): enables future delegation if needed
- $(document).keydown(...) → .on('keydown', ...): consistent event model
- $('#scrollToTop').click(...) → .on('click', ...): unified pattern across UI components

Additionally:
- Fixed event trigger in keyboard shortcut handler: now uses .trigger('focus') instead of direct .focus() to ensure consistent event flow
- Preserved event argument (e) and logic (e.preventDefault(), timeStamp checks) — no behavioral changes
- All selectors and functionality remain intact; this is a syntactic and maintainability upgrade

This refactor was guided by warnings from jQuery Migrate and is part of the larger migration plan:
1. Add jQuery Migrate → done
2. Fix deprecated patterns (like .bind(), .click()) → this commit
3. Upgrade to jQuery 3.7.1
4. Remove Migrate when clean

The code remains functionally identical, but is now more maintainable,
consistent, and future-proof. No regressions expected.

See: https://api.jquery.com/on/
See: https://github.com/jquery/jquery-migrate#readme

RafaelJohn9 committed:
chore: advance jQuery migration by upgrading to 3.7.1 with modern Migrate

Move to the latest stable jQuery (3.7.1) and jQuery UI (1.14.1), replacing outdated 1.9.1 versions. This continues our incremental modernization path.

Update jQuery Migrate to 3.4.1 to match, ensuring accurate deprecation warnings during development. The new Migrate version supports the latest APIs and provides better debugging feedback.

All existing behavior is preserved. Warnings in the console should now reflect any remaining legacy patterns that need cleanup.

Part of the migration sequence:
- Previous: refactor .bind()/.click() → .on()
- This: upgrade core libraries
- Next: remove Migrate once clean

RafaelJohn9 committed:
refactor: use .on() for keyup event in search input handler

Update the search input's  binding to use  instead of the deprecated shorthand , aligning with our consistent event binding pattern.

This change keeps the codebase uniform with recent refactors (e.g., click, focus, keydown) and prepares for safe compatibility with jQuery 3.x now that we've upgraded.

No behavioral changes event logic and timing remain identical.

Part of ongoing effort to eliminate legacy jQuery patterns ahead of removing jQuery Migrate.

RafaelJohn9 committed:
chore: remove jQuery Migrate after successful modernization

Remove jQuery Migrate 3.4.1 as it is no longer needed. The codebase has been fully updated to use modern jQuery patterns (e.g., .on() instead of .bind()), and no deprecation warnings have appeared in the console.

This marks the completion of our migration:
1. Added Migrate to detect issues
2. Refactored legacy event bindings
3. Upgraded to jQuery 3.7.1
4. Removed Migrate — now

RafaelJohn9 committed:
chore: remove outdated Sprockets require directives for jQuery

Clean up obsolete  comments in application.js that referenced old, externally loaded jQuery files:
- //= require jquery-1.7.1.min
- //= require jquery-ui-1.8.18.custom.min

These files are no longer part of the asset pipeline and are instead loaded directly via CDN in the layout. Leaving these directives was misleading and could cause confusion for future contributors.

This change aligns the comments with the current architecture: jQuery is now a standalone dependency (3.7.1 via CDN), not a compiled asset.

No functional impact — this is purely a documentation and hygiene improvement.

dscho reviewed (approved):
The commits are now in a great shape!

To1ne commented:
@RafaelJohn9 @dscho Awesome job! Thank you so much!
